### ISSUE TITLE
`jnp.repeat` doesn't lower to `jnp.full` without an explicit axis

### ISSUE THREAD

**Issue by jlebar** [google/jax#4099]
STR:

Start a JAX interpreter with envvars `CUDA_VISIBLE_DEVICES=0 XLA_FLAGS='--xla_dump_to=- --xla_dump_hlo_as_text'` and enable omnistaging (probably not necessary, but makes the logs a little less noisy): `import jax; jax.config.enable_omnistaging()`

If we `jnp.repeat` with an explicit `axis=0`, the HLO is `broadcast(-1)`, as desired:

```
> jax.jit(lambda a, repeats: jnp.repeat(a, repeats, axis=0), static_argnums=[0,1])(-1, jnp.asarray([10]))
*** Begin 1597811412237518.module_0002.before_optimizations.txt ***
HloModule jit__lambda___2.5

ENTRY %jit__lambda___2.5 () -> (s32[10]) {
  %constant.1 = pred[] constant(false)
  %constant.2 = s32[] constant(-1), metadata={op_type="broadcast_in_dim" op_name="jit(<lambda>)/broadcast_in_dim[ broadcast_dimensions=(  )\n                                shape=(10,) ]" source_file="<ipython-input-8-a49d7d47013d>" source_line=1}
  %broadcast.3 = s32[10]{0} broadcast(s32[] %constant.2), dimensions={}, metadata={op_type="broadcast_in_dim" op_name="jit(<lambda>)/broadcast_in_dim[ broadcast_dimensions=(  )\n                                shape=(10,) ]" source_file="<ipython-input-8-a49d7d47013d>" source_line=1}
  ROOT %tuple.4 = (s32[10]{0}) tuple(s32[10]{0} %broadcast.3)
}
*** End 1597811412237518.module_0002.before_optimizations.txt ***
```

But if we leave off the `axis=0` param, the generated HLO is a much more complicated way of expressing "all -1s".  I'm only going to post the after-optimization HLO, because before-optimization is quite large.

```
> jax.jit(lambda a, repeats: jnp.repeat(a, repeats), static_argnums=[0,1])(-1, jnp.asarray([10]))

*** Begin 1597811497850648.module_0003.after_optimizations.txt ***
HloModule jit__lambda___3.164

%fused_computation () -> s32[10] {
  %constant_223 = s32[1]{0} constant({-1}), metadata={op_type="reshape" op_name="jit(<lambda>)/reshape[ dimensions=None\n                       new_sizes=(1,) ]" source_file="<ipython-input-9-3872efff285b>" source_line=1}
  %constant_222 = s32[] constant(0), metadata={op_type="sub" op_name="jit(<lambda>)/sub" source_file="<ipython-input-9-3872efff285b>" source_line=1}
  %broadcast.32 = s32[10,1]{1,0} broadcast(s32[] %constant_222), dimensions={}, metadata={op_type="broadcast_in_dim" op_name="jit(<lambda>)/broadcast_in_dim[ broadcast_dimensions=(0,)\n                                shape=(10, 1) ]" source_file="<ipython-input-9-3872efff285b>" source_line=1}
  ROOT %gather.0 = s32[10]{0} gather(s32[1]{0} %constant_223, s32[10,1]{1,0} %broadcast.32), offset_dims={}, collapsed_slice_dims={0}, start_index_map={0}, index_vector_dim=1, slice_sizes={1}, metadata={op_type="gather" op_name="jit(<lambda>)/gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))\n                      slice_sizes=(1,) ]" source_file="<ipython-input-9-3872efff285b>" source_line=1}
}

ENTRY %jit__lambda___3.164 () -> (s32[10]) {
  %fusion = s32[10]{0} fusion(), kind=kLoop, calls=%fused_computation, metadata={op_type="gather" op_name="jit(<lambda>)/gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))\n                      slice_sizes=(1,) ]" source_file="<ipython-input-9-3872efff285b>" source_line=1}
  ROOT %tuple.163 = (s32[10]{0}) tuple(s32[10]{0} %fusion)
}
*** End 1597811497850648.module_0003.after_optimizations.txt ***
```

It looks like the implementation of `jnp.repeat` in lax_numpy actually tries to lower to `jnp.full`, but the code only works when axis is explicit.

This is not an important issue for me because it's easy to work around (supply the axis param, or just use jnp.full).  Maybe it's a good first bug for someone new to the project.

**Comment by parshnt** [comment]
I would like to understand JAX and contribute. Can I work on this?

**Comment by jakevdp** [comment]
@parshnt – thanks for your interest!

I looked into this, and it appears that it's in the process of being addressed at a deeper level, within XLA itself. So I think this issue can be closed.

If you're interested in getting started contributing to JAX, there is still some low-hanging fruit in issue #70.
